File: dotnet/Arkane/GameX.Arkane/Formats/PakBinary_Danae.cs
using GameX.Formats;
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;
using System.Threading.Tasks;

namespace GameX.Arkane.Formats
{
    public unsafe class PakBinary_Danae : PakBinary<PakBinary_Danae>
    {
        public override Task Read(BinaryPakFile source, BinaryReader r, object tag)
        {
            var files = source.Files = new List<FileSource>();
            var key = Encoding.ASCII.GetBytes((string)source.Game.Key); int keyLength = key.Length, keyIndex = 0;

            int readInt32(ref byte* b)
            {
                var p = b;
                *(p + 0) = (byte)(*(p + 0) ^ key[keyIndex++]); if (keyIndex >= keyLength) keyIndex = 0;
                *(p + 1) = (byte)(*(p + 1) ^ key[keyIndex++]); if (keyIndex >= keyLength) keyIndex = 0;
                *(p + 2) = (byte)(*(p + 2) ^ key[keyIndex++]); if (keyIndex >= keyLength) keyIndex = 0;
                *(p + 3) = (byte)(*(p + 3) ^ key[keyIndex++]); if (keyIndex >= keyLength) keyIndex = 0;
                b += 4;
                return *(int*)p;
            }

            string readString(ref byte* b)
            {
                var p = b;
                while (true)
                {
                    *p = (byte)(*p ^ key[keyIndex++]); if (keyIndex >= keyLength) keyIndex = 0;
                    if (*p == 0) break;
                    p++;
                }
                var length = (int)(p - b);
                var r = Encoding.ASCII.GetString(new ReadOnlySpan<byte>(b, length));
                b = p + 1;
                return r;
            }

            // move to fat table
            r.Seek(r.ReadUInt32());
            var fatSize = (int)r.ReadUInt32();
            var fatBytes = r.ReadBytes(fatSize);

            fixed (byte* _ = fatBytes)
            {
                byte* c = _, end = _ + fatSize;
                while (c < end)
                {
                    var dirPath = readString(ref c).Replace('\\', '/');
                    var numFiles = readInt32(ref c);
                    for (var i = 0; i < numFiles; i++)
                    {
                        var file = new FileSource
                        {
                            Path = dirPath + readString(ref c).Replace('\\', '/'),
                            Offset = readInt32(ref c),
                            Compressed = readInt32(ref c),
                            FileSize = readInt32(ref c),
                            PackedSize = readInt32(ref c),
                        };
                        if (file.Path.EndsWith(".FTL")) file.Compressed = 1;
                        else if (file.Compressed == 0) file.FileSize = file.PackedSize;
                        files.Add(file);
                    }
                }
            }
            return Task.CompletedTask;
        }

        public override Task<Stream> ReadData(BinaryPakFile source, BinaryReader r, FileSource file, FileOption option = default)
        {
            r.Seek(file.Offset);
            return Task.FromResult((Stream)new MemoryStream((file.Compressed & 1) != 0
                ? r.DecompressBlast((int)file.PackedSize, (int)file.FileSize)
                : r.ReadBytes((int)file.PackedSize)));
        }
    }
}

File: Python/gamex/Arkane/pakbinary_danae.py
import os
from io import BytesIO
from gamex.filesrc import FileSource
from gamex.pak import PakBinaryT
from gamex.compression import decompressBlast

# typedefs
class Reader: pass
class BinaryPakFile: pass

# PakBinary_Danae
class PakBinary_Danae(PakBinaryT):

    # read
    def read(self, source: BinaryPakFile, r: Reader, tag: object = None) -> None:
        source.files = files = []
        key = source.game.key.encode('ascii'); keyLength = len(key); keyIndex = 0

        # move to fat table
        r.seek(r.readUInt32())
        fatSize = r.readUInt32()
        fatBytes = bytearray(r.read(fatSize)); b = 0

        # read int32
        def readInt32() -> int:
            nonlocal b, keyIndex
            p = b
            fatBytes[p + 0] = fatBytes[p + 0] ^ key[keyIndex]; keyIndex += 1
            if keyIndex >= keyLength: keyIndex = 0
            fatBytes[p + 1] = fatBytes[p + 1] ^ key[keyIndex]; keyIndex += 1
            if keyIndex >= keyLength: keyIndex = 0
            fatBytes[p + 2] = fatBytes[p + 2] ^ key[keyIndex]; keyIndex += 1
            if keyIndex >= keyLength: keyIndex = 0
            fatBytes[p + 3] = fatBytes[p + 3] ^ key[keyIndex]; keyIndex += 1
            if keyIndex >= keyLength: keyIndex = 0
            b += 4
            return int.from_bytes(fatBytes[p:p+4], 'little', signed=True)

        # read string
        def readString() -> str:
            nonlocal b, keyIndex
            p = b
            while True:
                fatBytes[p] = fatBytes[p] ^ key[keyIndex]; keyIndex += 1
                if keyIndex >= keyLength: keyIndex = 0
                if fatBytes[p] == 0: break
                p += 1
            length = p - b
            r = fatBytes[b:p].decode('ascii', 'replace')
            b = p + 1
            return r

        # while there are bytes
        while b < fatSize:
            dirPath = readString().replace('\\', '/')
            numFiles = readInt32()
            for _ in range(numFiles):
                # get file
                file = FileSource(
                    path = dirPath + readString().replace('\\', '/'),
                    offset = readInt32(),
                    compressed = readInt32(),
                    fileSize = readInt32(),
                    packedSize = readInt32()
                    )
                # special case
                if file.path.endswith('.FTL'): file.compressed = 1
                elif file.compressed == 0: file.fileSize = file.packedSize
                # add file
                files.append(file)

    # readData
    def readData(self, source: BinaryPakFile, r: Reader, file: FileSource) -> BytesIO:
        r.seek(file.offset)
        return BytesIO(
            decompressBlast(r, file.packedSize, file.fileSize) if (file.compressed & 1) != 0 else \
            r.read(file.packedSize)
            )